pkg/cache/events.go

```diff
@@ -0,0 +1,34 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"fmt"
+	"sigs.k8s.io/karpenter/pkg/apis/v1beta1"
+
+	v1 "k8s.io/api/core/v1"
+
+	"sigs.k8s.io/karpenter/pkg/events"
+)
+
+func UnavailableOfferingEvent(nodeClaim *v1beta1.NodeClaim, instanceType, availabilityZone, capacityType string) events.Event {
+	return events.Event{
+		InvolvedObject: nodeClaim,
+		Type:           v1.EventTypeWarning,
+		Reason:         "UnavailableOffering",
+		Message:        fmt.Sprintf(`UnavailableOffering for {"instanceType": %q, "availabilityZone": %q, "capacityType": %q}`, instanceType, availabilityZone, capacityType),
+		DedupeValues:   []string{instanceType, availabilityZone, capacityType},
```

Comment:
We'd probably also want to rate limit this type of event generally. This is going to be really expensive to fire these events every time we get a spot insufficient capacity error, especially if they come in bursts, which I would generally expect them to.